crates/core_simd/src/ops.rs

```diff
@@ -0,0 +1,518 @@
+        impl_op! { impl Mul< <$type as crate::Vector>::Scalar > for $type }
+    };
+    { impl Div for $type:ty } => {
+        impl_op! { impl Div<$type> for $type }
+        impl_op! { impl Div< <$type as crate::Vector>::Scalar > for $type }
+    };
+    { impl Rem for $type:ty } => {
+        impl_op! { impl Rem<$type> for $type }
+        impl_op! { impl Rem< <$type as crate::Vector>::Scalar > for $type }
+    };
+    { impl BitAnd for $type:ty } => {
+        impl_op! { impl BitAnd<$type> for $type }
+        impl_op! { impl BitAnd< <$type as crate::Vector>::Scalar > for $type }
+    };
+    { impl BitOr for $type:ty } => {
+        impl_op! { impl BitOr<$type> for $type }
+        impl_op! { impl BitOr< <$type as crate::Vector>::Scalar > for $type }
+    };
+    { impl BitXor for $type:ty } => {
+        impl_op! { impl BitXor<$type> for $type }
+        impl_op! { impl BitXor< <$type as crate::Vector>::Scalar > for $type }
+    };
+
+    { impl Add<$rhs:ty> for $type:ty } => {
+        impl_op! { @binary $type, $rhs, Add::add, AddAssign::add_assign, simd_add }
+    };
+    { impl Sub<$rhs:ty> for $type:ty } => {
+        impl_op! { @binary $type, $rhs, Sub::sub, SubAssign::sub_assign, simd_sub }
+    };
+    { impl Mul<$rhs:ty> for $type:ty } => {
+        impl_op! { @binary $type, $rhs, Mul::mul, MulAssign::mul_assign, simd_mul }
+    };
+    { impl Div<$rhs:ty> for $type:ty } => {
+        impl_op! { @binary $type, $rhs, Div::div, DivAssign::div_assign, simd_div }
+    };
+    { impl Rem<$rhs:ty> for $type:ty } => {
+        impl_op! { @binary $type, $rhs, Rem::rem, RemAssign::rem_assign, simd_rem }
+    };
+    { impl Shl<$rhs:ty> for $type:ty } => {
+        impl_op! { @binary $type, $rhs, Shl::shl, ShlAssign::shl_assign, simd_shl }
+    };
+    { impl Shr<$rhs:ty> for $type:ty } => {
+        impl_op! { @binary $type, $rhs, Shr::shr, ShrAssign::shr_assign, simd_shr }
+    };
+    { impl BitAnd<$rhs:ty> for $type:ty } => {
+        impl_op! { @binary $type, $rhs, BitAnd::bitand, BitAndAssign::bitand_assign, simd_and }
+    };
+    { impl BitOr<$rhs:ty> for $type:ty } => {
+        impl_op! { @binary $type, $rhs, BitOr::bitor, BitOrAssign::bitor_assign, simd_or }
+    };
+    { impl BitXor<$rhs:ty> for $type:ty } => {
+        impl_op! { @binary $type, $rhs, BitXor::bitxor, BitXorAssign::bitxor_assign, simd_xor }
+    };
+
+    // Neg and Not implementations
+    { impl Not for $type:ty } => {
+        impl_ref_ops! {
+            impl core::ops::Not for $type {
+                type Output = Self;
+                fn not(self) -> Self::Output {
+                    self ^ <$type>::splat(true.into())
+                }
+            }
+        }
+    };
+
+    { impl Neg for $type:ty } => {
+        impl_ref_ops! {
+            impl core::ops::Neg for $type {
+                type Output = Self;
+                fn neg(self) -> Self::Output {
+                    self * <$type>::splat(-1 as <$type as crate::Vector>::Scalar)
```

Comment:
I normally do `0 - val` for Neg, since Sub is usually cheaper than Mul.

Would that give wrong values for any of the types?

Reply:
I wasn't sure how to implement `Neg` and this is what `packed-simd` does.  I assume that in reality both would get optimized to an actual neg op (like an xor or something?).  Maybe this optimizes better? Or maybe it doesn't matter at all?

Reply:
Let's put a TODO to check on it more carefully later on.

Reply:
I actually just changed it to use `0 - val` since it was much simpler to implement (`$type::default - self`)